src/contracts/StakingChef.sol

```diff
@@ -0,0 +1,184 @@
+            rewardDebt: metricAmount.mul(_lifetimeShareValue).div(ACC_METRIC_PRECISION),
+            claimable: 0
+        });
+
+        _stakes.push(stake);
+        _totalAllocPoint = _totalAllocPoint.add(stake.metricAmount);
+        SafeERC20.safeTransferFrom(IERC20(metric), msg.sender, address(this), stake.metricAmount); 
+    }
+
+    function updateStaker(
+        address stakeAddress,
+        uint256 stakeIndex,
+        uint256 metricAmount 
+    ) public {
+        if (areRewardsActive() && _totalAllocPoint > 0) {
+            updateAccumulatedStakingRewards();
+        }
+        _totalAllocPoint = _totalAllocPoint.sub(_stakes[stakeIndex].metricAmount).add(metricAmount);
+        _stakes[stakeIndex].stakeAddress = stakeAddress;
+        _stakes[stakeIndex].metricAmount = metricAmount;
+    }
+
+    function removeStaker(uint256 stakeIndex) external {
+        require(stakeIndex < _stakes.length);
+        if (areRewardsActive() && _totalAllocPoint > 0) {
+            updateAccumulatedStakingRewards();
+        }
+        _totalAllocPoint = _totalAllocPoint.sub(_stakes[stakeIndex].metricAmount);
+
+        _stakes[stakeIndex] = _stakes[_stakes.length - 1];
+        _stakes.pop();
+    }
+
+    function stakeAdditionalMetric(
+        address stakeAddress,
+        uint256 stakeIndex,
+        uint256 metricAmount,
+        uint256 newStartDate
+    ) public {
+        harvest(stakeIndex);
+        uint256 principalMetric = _stakes[stakeIndex].metricAmount;
+        uint256 totalMetricStaked = metricAmount + principalMetric;
+
+        Staker memory stake = Staker({
+            stakeAddress: stakeAddress,
+            metricAmount: totalMetricStaked,
+            startDate: newStartDate,
+            rewardDebt: metricAmount.mul(_lifetimeShareValue).div(ACC_METRIC_PRECISION),
+            claimable: 0
+            });
+
+            _stakes[stakeIndex] = stake;
+            _totalAllocPoint = _totalAllocPoint.add(stake.metricAmount);
+             SafeERC20.safeTransferFrom(IERC20(metric), msg.sender, address(this), stake.metricAmount);
+    }
+
+    function updateAccumulatedStakingRewards() public {
+        require(areRewardsActive(), "Rewards are not active");
+        if (block.number <= getLastRewardBlock()) {
+            return;
+        }
+
+        uint256 blocks = block.number.sub(getLastRewardBlock());
+
+        uint256 accumulated = blocks.mul(getMetricPerBlock());
+
+        _lifetimeShareValue = _lifetimeShareValue.add(accumulated.mul(ACC_METRIC_PRECISION).div(_totalAllocPoint));
+        setLastRewardBlock(block.number);
+    }
+
+// --------------------------------------------------------------------- Manage rewards and Principal
+
+    function claim(uint256 stakeIndex) public {
+        Staker storage stake = _stakes[stakeIndex];
+        harvest(stakeIndex);
+
+        require(stake.claimable != 0, "No claimable rewards to withdraw");
+        require(stake.stakeAddress == _msgSender(), "Sender can not claim");
```

Comment:
require message is inverted -> "Only staker can claim?" 

But also when we switch from an array to a mapping this won't be a problem, so probably no biggie